collector/perf_linux.go

```diff
@@ -0,0 +1,549 @@
+// Copyright 2019 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"fmt"
+	"runtime"
+	"sync"
+
+	perf "github.com/hodgesds/perf-utils"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+)
+
+const (
+	perfSubsystem = "perf"
+)
+
+var (
+	perfHwProfilers    = map[int]perf.HardwareProfiler{}
+	perfSwProfilers    = map[int]perf.SoftwareProfiler{}
+	perfCacheProfilers = map[int]perf.CacheProfiler{}
+	perfInitializeOnce sync.Once
+)
+
+func init() {
+	registerCollector(perfSubsystem, defaultDisabled, NewPerfCollector)
+}
+
+// perfCollector is a Collected that uses the perf subsystem to collect
+// metrics. It uses perf_event_open an ioctls for profiling.
+type perfCollector struct{}
+
+// initializePerfProfilers is used to setup
+func initializePerfProfilers() {
+	ncpus := runtime.NumCPU()
+	for i := 0; i < ncpus; i++ {
+		// Use -1 to profile all processes on the CPU, see:
+		// man perf_event_open
+		perfHwProfilers[i] = perf.NewHardwareProfiler(-1, i)
+		if err := perfHwProfilers[i].Start(); err != nil {
```

Comment:
If you move this to NewPerfCollector, we can just return the error instead of logging and continuing.

Reply:
Tested this out this morning and that seems to work out nice when no profilers are available.